tests/test_qnode.py

```diff
@@ -14,377 +14,553 @@
+            (64,),
+            (64, 1),
+            (32, 2),
+            (16, 4),
+            (8, 8),
+            (16, 2, 2),
+            (8, 2, 2, 2),
+            (4, 2, 2, 2, 2),
+            (2, 2, 2, 2, 2, 2),
+        ],
+    )
+    def test_flatten(self, shape):
+        """Tests that _flatten successfully flattens multidimensional arrays."""
+
+        reshaped = np.reshape(flat_dummy_array, shape)
+        flattened = np.array([x for x in _flatten(reshaped)])
+
+        assert flattened.shape == flat_dummy_array.shape
+        assert np.array_equal(flattened, flat_dummy_array)
+
+    @pytest.mark.parametrize(
+        "shape",
+        [
+            (64,),
+            (64, 1),
+            (32, 2),
+            (16, 4),
+            (8, 8),
+            (16, 2, 2),
+            (8, 2, 2, 2),
+            (4, 2, 2, 2, 2),
+            (2, 2, 2, 2, 2, 2),
+        ],
+    )
+    def test_unflatten(self, shape):
+        """Tests that _unflatten successfully unflattens multidimensional arrays."""
+
+        reshaped = np.reshape(flat_dummy_array, shape)
+        unflattened = np.array([x for x in unflatten(flat_dummy_array, reshaped)])
+
+        assert unflattened.shape == reshaped.shape
+        assert np.array_equal(unflattened, reshaped)
+
+    def test_unflatten_error_unsupported_model(self):
+        """Tests that unflatten raises an error if the given model is not supported"""
+
+        with pytest.raises(TypeError, match="Unsupported type in the model"):
+            model = lambda x: x  # not a valid model for unflatten
+            unflatten(flat_dummy_array, model)
 
-from pennylane.qnode import _flatten, unflatten, QNode, QuantumFunctionError
-from pennylane.plugins.default_qubit import CNOT, Rotx, Roty, Rotz, I, CRotx, CRoty, CRotz, Y, Z
-from pennylane._device import DeviceError
+    def test_unflatten_error_too_many_elements(self):
+        """Tests that unflatten raises an error if the given iterable has
+           more elements than the model"""
 
+        reshaped = np.reshape(flat_dummy_array, (16, 2, 2))
 
-def expZ(state):
-    return np.abs(state[0]) ** 2 - np.abs(state[1]) ** 2
+        with pytest.raises(ValueError, match="Flattened iterable has more elements than the model"):
+            unflatten(np.concatenate([flat_dummy_array, flat_dummy_array]), reshaped)
 
 
-thetas = np.linspace(-2*np.pi, 2*np.pi, 7)
+@pytest.fixture(scope="function")
+def opqueue_test_node(mock_device):
+    """Provides a circuit for the subsequent tests of the operation queue"""
 
-a = np.linspace(-1,1,64)
-a_shapes = [(64,),
-            (64,1),
-            (32,2),
-            (16,4),
-            (8,8),
-            (16,2,2),
-            (8,2,2,2),
-            (4,2,2,2,2),
-            (2,2,2,2,2,2)]
+    def circuit(x):
+        qml.RX(x, wires=[0])
+        qml.CNOT(wires=[0, 1])
+        qml.RY(0.4, wires=[0])
+        qml.RZ(-0.2, wires=[1])
+        return qml.expval(qml.PauliX(0)), qml.expval(qml.PauliZ(1))
```

Comment:
If a fixture is only being used in a single class of tests, lets make it a method of that class, i.e.,

```python
class TestQNodeOperationQueue:
    """Tests that the QNode operation queue is properly filled and interacted with"""

    @pytest.fixture(scope="function")
    def opqueue_test_node(self, mock_device):
        """Provides a circuit for the subsequent tests of the operation queue"""

        def circuit(x):
            qml.RX(x, wires=[0])
            qml.CNOT(wires=[0, 1])
            qml.RY(0.4, wires=[0])
            qml.RZ(-0.2, wires=[1])
            return qml.expval(qml.PauliX(0)), qml.expval(qml.PauliZ(1))

        node = qml.QNode(circuit, mock_device)
        node.construct([1.0])

        return node
```

More of a stylistic than a functional change, but it helps take advantage of Python's indentation to help visually group fixtures with test functions.

Reply:
Actually, I didn't know one can do that! I agree that this is the better style.